solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java

```diff
@@ -49,411 +58,673 @@
-    } finally {
-      if (leaderDirectClient != null) {
-        leaderDirectClient.close();
-      }
-      if (followerDirectClient != null) {
-        followerDirectClient.close();
-      }
-      if (core != null) {
-        core.close();
-      }
-      if (cluster != null) {
-        cluster.shutdown();
+  /**
+   * Goes over all the lives of a node(node gets a new life on restart) and then goes over each core's concurrency stages
+   * in each life. Logs the concurrency stages in the order they occurred and then analyze those stages to make sure no
+   * critical section was breached.
+   */
+  private void analyzeCoreConcurrencyStagesForBreaches() {
+    // Goes over each node
+    for (Map.Entry<String, List<SolrProcessTracker>> nodeTracker :
+        testState.solrNodesTracker.entrySet()) {
+      String nodeName = nodeTracker.getKey();
+      int lifeCountForNode = nodeTracker.getValue().size();
+      // Goes over each life of a node
+      for (int i = 0; i < lifeCountForNode; i++) {
+        ConcurrentHashMap<String, ConcurrentLinkedQueue<String>> coreConcurrencyStageTracker = nodeTracker.getValue().get(i).coreConcurrencyStageTracker;
+        if (coreConcurrencyStageTracker.isEmpty()) {
+          log.info("life " + (i + 1) + "/" + lifeCountForNode + " of node " + nodeName + " is empty");
+        } else {
+          // Goes over each core
+          for (Map.Entry<String, ConcurrentLinkedQueue<String>> coreConcurrencyStagesEntry : coreConcurrencyStageTracker.entrySet()) {
+            String coreName = coreConcurrencyStagesEntry.getKey();
+            List<String> coreConcurrencyStages = new ArrayList<>(coreConcurrencyStagesEntry.getValue());
+            // Log line is truncated beyond certain length, therefore, printing them in the batches of 200
+            List<List<String>> batches = Lists.partition(coreConcurrencyStages, 200);
+            if (batches.isEmpty()) {
+              batches = new ArrayList<>(1);
+              batches.add(new ArrayList<>(0));
+            }
+            for (int j = 0; j < batches.size(); j++) {
+              log.info("batch " + (j + 1) + "/" + batches.size()
+                  + " of core " + coreName
+                  + " of life " + (i + 1) + "/" + lifeCountForNode
+                  + " of node " + nodeName
+                  + "\n" + batches.get(j).toString());
+            }
+            analyzeCoreConcurrencyStagesForBreaches(coreName, coreConcurrencyStages);
+          }
+        }
       }
-      // clean up the shared store. The temp dir should clean up itself after the test class finishes
-      FileUtils.cleanDirectory(sharedStoreRootPath.toFile());
     }
   }
 
-  private void assertCriticalSections(List<String> progress) {
-    String currentThreadId = null;
+  /**
+   * Analyze core's concurrency stages to make sure no critical section was breached.
+   */
+  private void analyzeCoreConcurrencyStagesForBreaches(String coreName, List<String> coreConcurrencyStages) {
```

Comment:
A note pointing to SharedCoreConcurrencyController here might help future devs understand the implied contracts we're testing here

Reply:
Updated comment.